pkg/cloudprovider/provider/packet/provider.go

```diff
@@ -0,0 +1,430 @@
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c.APIKey)
+
+	imageName, err := getNameForOS(pc.OperatingSystem)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Invalid operating system specified %q, details = %v", pc.OperatingSystem, err),
+		}
+	}
+
+	serverCreateOpts := &packngo.DeviceCreateRequest{
+		Hostname:  machine.Spec.Name,
+		UserData:  userdata,
+		ProjectID: c.ProjectID,
+		Facility:  c.Facilities,
+		Plan:      c.InstanceType,
+		OS:        imageName,
+		Tags: []string{
+			generateTag(string(machine.UID)),
+		},
+	}
+
+	device, res, err := client.Devices.Create(serverCreateOpts)
+	if err != nil {
+		return nil, packetErrorToTerminalError(err, res, "failed to create server")
+	}
+
+	return &packetDevice{device: device}, nil
+}
+
+func (p *provider) Cleanup(machine *v1alpha1.Machine, _ *cloud.MachineCreateDeleteData) (bool, error) {
+	instance, err := p.Get(machine)
+	if err != nil {
+		if err == cloudprovidererrors.ErrInstanceNotFound {
+			return true, nil
+		}
+		return false, err
+	}
+
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return false, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c.APIKey)
+
+	res, err := client.Devices.Delete(instance.(*packetDevice).device.ID)
+	if err != nil {
+		return false, packetErrorToTerminalError(err, res, "failed to delete the server")
+	}
+
+	return false, nil
+}
+
+func (p *provider) AddDefaults(spec v1alpha1.MachineSpec) (v1alpha1.MachineSpec, error) {
+	return spec, nil
+}
+
+func (p *provider) Get(machine *v1alpha1.Machine) (instance.Instance, error) {
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c.APIKey)
+	device, err := getDeviceByTag(client, c.ProjectID, generateTag(string(machine.UID)))
+	if err != nil {
+		return nil, err
+	}
+
+	if device != nil {
+		return &packetDevice{device: device}, nil
+	}
+
+	return nil, cloudprovidererrors.ErrInstanceNotFound
+}
+
+func (p *provider) MigrateUID(machine *v1alpha1.Machine, newID types.UID) error {
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	tag := generateTag(string(machine.UID))
+	client := getClient(c.APIKey)
+	device, err := getDeviceByTag(client, c.ProjectID, tag)
+	if err != nil {
+		return err
+	}
+
+	// append to our tags
+	tags := append(device.Tags, tag)
```

Comment:
We need to actually replace the old one, ideally with one API call so we cant end up with both a new and the old tag

Reply:
Can you explain it some more? What are we looking to do?

Reply:
We use the UID of the Kubernetes object as unique identifier of the object in the providers API. That means the object in the providers API must have exactly one UID tag/label whatever it offers. If I don't misunderstand your code, it will currently put a second UID tag on the instance when `MigrateUID` gets called

Reply:
Tags at packet aren't key/value pairs, but just arbitrary strings. You easily can make it key/value by making the string, e.g., 

```yml
- "Name:Alvaro"
- "Purpose:Success"
```

etc.

In this case, it might have other tags, perhaps even unrelated to this specific UID tag. So I am appending this new tag to the existing tags, whatever they may be.

However, I see how you might get two or more actual UID tags (independent of whatever other tags there are). I am going to change it so that it ensures you only get one of the UID tags.

Reply:
OK done.